File: crates/5kill5-core/src/quality.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Severity {
    Blocker,
    High,
    Medium,
    Low,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Finding {
    pub severity: Severity,
    pub rule_id: String,
    pub line_start: usize,
    pub line_end: usize,
    pub message: String,
    pub recommendation: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QualityReport {
    pub status: String,
    pub score: u8,
    pub findings: Vec<Finding>,
}

pub fn scan_skill(content: &str) -> QualityReport {
    let frontmatter = parse_frontmatter(content);
    let mut findings = Vec::new();

    match frontmatter.get("name") {
        Some(value) if !value.trim().is_empty() => {}
        _ => findings.push(finding(
            Severity::Blocker,
            "format.required_frontmatter",
            1,
            "Missing required manifest field: name",
            "Add name to YAML front matter.",
        )),
    }

    match frontmatter.get("version") {
        Some(value) if is_semver(value) => {}
        Some(value) => findings.push(finding(
            Severity::Blocker,
            "format.semver",
            1,
            &format!("Invalid SemVer: {value}"),
            "Use a version such as 1.0.0.",
        )),
        None => findings.push(finding(
            Severity::Blocker,
            "format.required_frontmatter",
            1,
            "Missing required manifest field: version",
            "Add version to YAML front matter.",
        )),
    }

    if !has_heading(content, "Trigger") || !has_heading(content, "Steps") {
        findings.push(finding(
            Severity::Blocker,
            "format.required_sections",
            1,
            "Required sections are incomplete",
            "Include ## Trigger and ## Steps sections.",
        ));
    }

    for (line_number, line) in content.lines().enumerate() {
        let line_start = line_number + 1;
        let lower = line.to_ascii_lowercase();

        if has_secret_pattern(line) {
            findings.push(finding(
                Severity::High,
                "security.secret_patterns",
                line_start,
                "Potential secret-like value found",
                "Review the evidence and remove secrets before upload.",
            ));
        }

        if has_destructive_command(&lower) {
            findings.push(finding(
                Severity::High,
                "security.destructive_commands",
                line_start,
                "Potential destructive command found",
                "Keep destructive actions out of skill instructions or require explicit review.",
            ));
        }

        if has_remote_exec(&lower) {
            findings.push(finding(
                Severity::High,
                "supply_chain.remote_fetch",
                line_start,
                "Remote script execution pattern found",
                "Replace pipe-to-shell installation with inspectable steps.",
            ));
        }

        if has_prompt_injection(&lower) {
            findings.push(finding(
                Severity::Medium,
                "security.prompt_injection",
                line_start,
                "Prompt-injection language found",
                "Review wording and allowlist only after manual verification.",
            ));
        }
    }

    let penalty: usize = findings
        .iter()
        .map(|item| match item.severity {
            Severity::Blocker => 45,
            Severity::High => 25,
            Severity::Medium => 12,
            Severity::Low => 5,
        })
        .sum();
    let status = if findings.iter().any(|item| item.severity == Severity::Blocker) {
        "blocked"
    } else if findings.is_empty() {
        "passed"
    } else {
        "warning"
    };

    QualityReport {
        status: status.to_string(),
        score: 100usize.saturating_sub(penalty) as u8,
        findings,
    }
}

fn finding(
    severity: Severity,
    rule_id: &str,
    line_start: usize,
    message: &str,
    recommendation: &str,
) -> Finding {
    Finding {
        severity,
        rule_id: rule_id.to_string(),
        line_start,
        line_end: line_start,
        message: message.to_string(),
        recommendation: recommendation.to_string(),
    }
}

fn parse_frontmatter(content: &str) -> std::collections::HashMap<String, String> {
    let mut values = std::collections::HashMap::new();
    let mut lines = content.lines();
    if lines.next().map(str::trim) != Some("---") {
        return values;
    }

    for line in lines {
        let trimmed = line.trim();
        if trimmed == "---" {
            break;
        }
        if let Some((key, value)) = trimmed.split_once(':') {
            if !value.trim().is_empty() {
                values.insert(key.trim().to_string(), value.trim().trim_matches('"').to_string());
            }
        }
    }

    values
}

fn is_semver(version: &str) -> bool {
    let core = version.split(['-', '+']).next().unwrap_or(version);
    let mut parts = core.split('.');
    let Some(major) = parts.next() else { return false };
    let Some(minor) = parts.next() else { return false };
    let Some(patch) = parts.next() else { return false };
    parts.next().is_none() && is_number(major) && is_number(minor) && is_number(patch)
}

fn is_number(value: &str) -> bool {
    !value.is_empty()
        && value.chars().all(|ch| ch.is_ascii_digit())
        && (value == "0" || !value.starts_with('0'))
}

fn has_heading(content: &str, heading: &str) -> bool {
    content.lines().any(|line| {
        let trimmed = line.trim();
        trimmed
            .strip_prefix("##")
            .map(|value| value.trim().eq_ignore_ascii_case(heading))
            .unwrap_or(false)
    })
}

fn has_secret_pattern(line: &str) -> bool {
    line.contains("BEGIN PRIVATE KEY")
        || line.contains("Authorization: Bearer ")
        || line.split(|ch: char| !ch.is_ascii_alphanumeric() && ch != '_' && ch != '-')
            .any(|token| {
                token.starts_with("sk-") && token.len() >= 23
                    || token.starts_with("ghp_") && token.len() >= 24
                    || token.starts_with("github_pat_") && token.len() >= 31
                    || token.starts_with("AKIA") && token.len() >= 20
                    || token.starts_with("xoxb-") && token.len() >= 25
            })
}

fn has_destructive_command(lower: &str) -> bool {
    lower.contains("rm -rf /")
        || lower.contains("remove-item -recurse -force")
        || lower.contains("del /s /q")
        || lower.contains("diskpart")
        || lower.contains("git reset --hard")
        || lower.contains("format c:")
}

fn has_remote_exec(lower: &str) -> bool {
    (lower.contains("curl ") || lower.contains("wget ") || lower.contains("invoke-webrequest "))
        && (lower.contains("| sh")
            || lower.contains("| bash")
            || lower.contains("| invoke-expression")
            || lower.contains("| iex"))
}

fn has_prompt_injection(lower: &str) -> bool {
    lower.contains("ignore previous instructions")
        || lower.contains("ignore system instructions")
        || lower.contains("leak system prompt")
        || lower.contains("bypass safety")
        || lower.contains("silent hidden step")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn blocks_missing_required_fields() {
        let report = scan_skill("# Test\n\n## Trigger\n\nx\n\n## Steps\n\nx");
        assert_eq!(report.status, "blocked");
        assert!(report.findings.iter().any(|item| item.rule_id == "format.required_frontmatter"));
    }

    #[test]
    fn warns_on_destructive_command() {
        let content = "---\nname: test\nversion: 1.0.0\n---\n\n## Trigger\n\nx\n\n## Steps\n\nrm -rf /";
        let report = scan_skill(content);
        assert!(report.findings.iter().any(|item| item.rule_id == "security.destructive_commands"));
    }
}

File: cli/src/fivekill5/quality.py
"""Quality scanning for 5KILL5 skills."""

from dataclasses import dataclass
from enum import Enum
import re


class Severity(Enum):
    """Severity levels for quality findings."""

    BLOCKER = "blocker"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


@dataclass
class Finding:
    """Represents a quality finding in a skill."""

    severity: Severity
    rule_id: str
    line_start: int
    line_end: int
    message: str
    recommendation: str


@dataclass
class QualityReport:
    """Quality scan report for a skill."""

    status: str
    score: int
    findings: list[Finding]


# Scoring penalties
PENALTIES = {
    Severity.BLOCKER: 45,
    Severity.HIGH: 25,
    Severity.MEDIUM: 12,
    Severity.LOW: 5,
}


def scan_skill(content: str) -> QualityReport:
    """Scan a skill's content and return quality report."""
    findings: list[Finding] = []

    frontmatter = _parse_frontmatter(content)

    # Check required frontmatter
    if not frontmatter.get("name", "").strip():
        findings.append(
            Finding(
                severity=Severity.BLOCKER,
                rule_id="format.required_frontmatter",
                line_start=1,
                line_end=1,
                message="Missing required manifest field: name",
                recommendation="Add name to YAML front matter.",
            )
        )

    version = frontmatter.get("version", "")
    if not version:
        findings.append(
            Finding(
                severity=Severity.BLOCKER,
                rule_id="format.required_frontmatter",
                line_start=1,
                line_end=1,
                message="Missing required manifest field: version",
                recommendation="Add version to YAML front matter.",
            )
        )
    elif not _is_semver(version):
        findings.append(
            Finding(
                severity=Severity.BLOCKER,
                rule_id="format.semver",
                line_start=1,
                line_end=1,
                message=f"Invalid SemVer: {version}",
                recommendation="Use a version such as 1.0.0.",
            )
        )

    # Check required sections
    if not _has_heading(content, "Trigger") or not _has_heading(content, "Steps"):
        findings.append(
            Finding(
                severity=Severity.BLOCKER,
                rule_id="format.required_sections",
                line_start=1,
                line_end=1,
                message="Required sections are incomplete",
                recommendation="Include ## Trigger and ## Steps sections.",
            )
        )

    # Line-by-line security checks
    for line_num, line in enumerate(content.split("\n"), start=1):
        line_lower = line.lower()

        # Secret patterns
        if _has_secret_pattern(line):
            findings.append(
                Finding(
                    severity=Severity.HIGH,
                    rule_id="security.secret_patterns",
                    line_start=line_num,
                    line_end=line_num,
                    message="Potential secret-like value found",
                    recommendation="Review the evidence and remove secrets before upload.",
                )
            )

        # Destructive commands
        if _has_destructive_command(line_lower):
            findings.append(
                Finding(
                    severity=Severity.HIGH,
                    rule_id="security.destructive_commands",
                    line_start=line_num,
                    line_end=line_num,
                    message="Potential destructive command found",
                    recommendation="Keep destructive actions out of skill instructions or require explicit review.",
                )
            )

        # Remote script execution
        if _has_remote_exec(line_lower):
            findings.append(
                Finding(
                    severity=Severity.HIGH,
                    rule_id="supply_chain.remote_fetch",
                    line_start=line_num,
                    line_end=line_num,
                    message="Remote script execution pattern found",
                    recommendation="Replace pipe-to-shell installation with inspectable steps.",
                )
            )

        # Prompt injection
        if _has_prompt_injection(line_lower):
            findings.append(
                Finding(
                    severity=Severity.MEDIUM,
                    rule_id="security.prompt_injection",
                    line_start=line_num,
                    line_end=line_num,
                    message="Prompt-injection language found",
                    recommendation="Review wording and allowlist only after manual verification.",
                )
            )

    # Calculate score
    penalty = sum(PENALTIES[f.severity] for f in findings)
    score = max(0, 100 - penalty)

    # Determine status
    if any(f.severity == Severity.BLOCKER for f in findings):
        status = "blocked"
    elif not findings:
        status = "passed"
    else:
        status = "warning"

    return QualityReport(status=status, score=score, findings=findings)


def _parse_frontmatter(content: str) -> dict[str, str]:
    """Parse YAML frontmatter from skill content."""
    values: dict[str, str] = {}
    lines = content.split("\n")
    
    if not lines or lines[0].strip() != "---":
        return values

    for line in lines[1:]:
        trimmed = line.strip()
        if trimmed == "---":
            break
        if ":" in trimmed:
            key, value = trimmed.split(":", 1)
            value = value.strip().strip('"').strip("'")
            if value:
                values[key.strip()] = value

    return values


def _is_semver(version: str) -> bool:
    """Check if a version string is valid SemVer."""
    # Remove pre-release and build metadata
    core = version.split("-")[0].split("+")[0]
    parts = core.split(".")

    if len(parts) < 3:
        return False

    major, minor, patch = parts[0], parts[1], parts[2]
    
    # Check all parts are numeric
    if not (major.isdigit() and minor.isdigit() and patch.isdigit()):
        return False

    # No leading zeros (except for 0 itself)
    if major != "0" and major.startswith("0"):
        return False
    if minor != "0" and minor.startswith("0"):
        return False
    if patch != "0" and patch.startswith("0"):
        return False

    return True


def _has_heading(content: str, heading: str) -> bool:
    """Check if content contains a specific heading."""
    pattern = re.compile(r"^##\s+(.+)$", re.MULTILINE | re.IGNORECASE)
    for match in pattern.finditer(content):
        if match.group(1).strip().lower() == heading.lower():
            return True
    return False


def _has_secret_pattern(line: str) -> bool:
    """Check if a line contains potential secret patterns."""
    # Direct patterns
    if "BEGIN PRIVATE KEY" in line:
        return True
    if "Authorization: Bearer " in line:
        return True

    # Token patterns (sk-, ghp_, github_pat_, AKIA, xoxb-)
    patterns = [
        (r"\bsk-[a-zA-Z0-9]{20,}", 23),
        (r"\bghp_[a-zA-Z0-9]{20,}", 24),
        (r"\bgithub_pat_[a-zA-Z0-9_]{20,}", 31),
        (r"\bAKIA[A-Z0-9]{16,}", 20),
        (r"\bxoxb-[a-zA-Z0-9-]{20,}", 25),
    ]

    for pattern, min_len in patterns:
        if re.search(pattern, line):
            return True

    return False


def _has_destructive_command(lower: str) -> bool:
    """Check for destructive command patterns."""
    patterns = [
        r"rm\s+-rf\s+/",
        r"remove-item\s+-recurse\s+-force",
        r"del\s+/s\s+/q",
        r"diskpart",
        r"git\s+reset\s+--hard",
        r"format\s+c:",
    ]

    for pattern in patterns:
        if re.search(pattern, lower):
            return True

    return False


def _has_remote_exec(lower: str) -> bool:
    """Check for remote script execution patterns."""
    # Must have curl/wget/invoke-webrequest AND pipe to shell
    has_fetcher = bool(re.search(r"(curl|wget|invoke-webrequest)\s+", lower))
    has_pipe_shell = bool(
        re.search(r"\|\s*(sh|bash|invoke-expression|iex)", lower)
    )

    return has_fetcher and has_pipe_shell


def _has_prompt_injection(lower: str) -> bool:
    """Check for prompt injection patterns."""
    patterns = [
        r"ignore\s+(previous|all)\s+instructions",
        r"ignore\s+system\s+instructions",
        r"leak\s+system\s+prompt",
        r"bypass\s+safety",
        r"silent\s+hidden\s+step",
    ]

    for pattern in patterns:
        if re.search(pattern, lower):
            return True

    return False
